server/server.go

```diff
@@ -9,17 +9,67 @@ import (
 
 	"github.com/bombsimon/amqp-rpc/connection"
 	"github.com/bombsimon/amqp-rpc/logger"
-	"github.com/bombsimon/amqp-rpc/middleware"
+)
+
+type ctxKey string
+
+const (
+	// CtxQueueName can be used to get the queue name from the context.Context
+	// inside the HandlerFunc.
+	CtxQueueName ctxKey = "queue_name"
 )
 
 var (
 	// ErrResponseChClosed is an error representing a closed response channel.
 	ErrResponseChClosed = errors.New("Channel closed")
 )
 
-// handlerFunc is the function that handles all request based on the routing
+// HandlerFunc is the function that handles all request based on the routing
 // key.
-type handlerFunc func(context.Context, amqp.Delivery) []byte
+type HandlerFunc func(context.Context, *ResponseWriter, amqp.Delivery)
+
+// ResponseWriter is used by a handler to construct an RPC response.
+// The ResponseWriter may NOT be used after the handler has returned.
+type ResponseWriter struct {
+	publishing *amqp.Publishing
+	mandatory  bool
+	immediate  bool
+}
+
+// Write will write the response Body of the amqp.Publishing.
+// It is safe to call Write multiple times.
+func (rw *ResponseWriter) Write(p []byte) (int, error) {
+	rw.publishing.Body = append(rw.publishing.Body, p...)
+	return len(p), nil
+}
+
+// Publishing returns the internal amqp.Publishing that are used for the response,
+// useful for modification.
+func (rw *ResponseWriter) Publishing() *amqp.Publishing {
+	return rw.publishing
+}
+
+// Mandatory sets the mandatory flag on the later amqp.Publish.
+func (rw *ResponseWriter) Mandatory(m bool) {
+	rw.mandatory = m
+}
+
+// Immediate sets the immediate flag on the later amqp.Publish.
+func (rw *ResponseWriter) Immediate(i bool) {
+	rw.immediate = i
+}
+
+// processedRequest is used to add the response from a handler func combined
+// with a amqp.Delivery. The reasone we need to combine those is that we reply
+// to each request in a separate go routine and the delivery is required to
+// determine on which queue to reply.
+type processedRequest struct {
+	correlationID string
```

Comment:
Forgot to remove this? I can't see it's being used and the current correlationID is found in the `amqp.Publishing`.